near-sdk/src/environment/env.rs

```diff
@@ -149,13 +150,13 @@ pub fn register_len(register_id: u64) -> Option<u64> {
 // ###############
 /// The id of the account that owns the current contract.
 pub fn current_account_id() -> AccountId {
-    String::from_utf8(method_into_register!(current_account_id)).unwrap()
+    AccountId::try_from(method_into_register!(current_account_id)).unwrap()
 }
```

Comment:
hmm, the unreachable definitely feels uncertain to me, since this account id validation is custom to the SDK and if there ever was a discrepancy, this would be a tough error to debug. I will change to this now though

Reply:
Also, why do we want to remove `TryFrom<Vec<u8>>`? I would have thought the opposite with the trait being part of the core prelude with the next Rust edition

Reply:
>hmm, the unreachable definitely feels uncertain to me, since this account id validation is custom to the SDK and if there ever was a discrepancy, this would be a tough error to debug.

So, the `.unreachable` is more or less equivalent to `.unwrap()`, the difference is that it doesn't invoke formatting machinery. Unwrap is defined as

![image](https://user-images.githubusercontent.com/1711539/124582455-b16bff80-de5a-11eb-9885-eb96d0fc8f65.png)

and `E: Debug` is what's problematic.


>Also, why do we want to remove TryFrom<Vec<u8>>?

The problem with that is that it causes us to have `InvalidUtf8` as separate error condition in the ErrorKind enum. Today, that's just a field-less variant, but that's suboptimal, as there's a perfectly fine type for that in the standard library: https://doc.rust-lang.org/stable/std/string/struct.FromUtf8Error.html. Not that it allows recovering the original bytes, and also allows to know what exactly failed during utf8 validation (which byte is faulty). If we don't expose this functionality, we provide half-baked API. If we do expose this functionality, we increase API surface and design space a bit. 

A simpler solution is to let the caller handle "invalid ut8" and "invalid account id" errors separately: `String::from_utf8(bytes).and_then(AccountId::try_from)`. This gives the maximal flexibility to the user to handle utf8 errors however they like, and allows us to not think about this error case at all. One drawback here is that this is potentially less optimal -- in theory, you can validate both utf8 validity and account id validity in a single pass over data, but that's a very small benefit for high implementation complexity (and potentially some code bloat), so I don't think we'll ever do that. 


Separately, it seems that the main beneficiary of this impl is actually the SDK itself? It's SDK who gets raw bytes from the extern "C" functions. The contracts themselves seem to work with utf8 directly.